### PR TITLE
Doc: Clarify CentOS 7 kernel upgrade and Cilium advance features

### PR DESCRIPTION
**Pull request by vincentmli**
Users run CentOS 7 with third-party kernel upgrade with
    Cilium advance features should check various advance features
    kernel config options requirements.

Signed-off-by: Vincent Li <v.li@f5.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
doc: clarify CentOS 7 third-part kernel upgrade and Cilium advance features kernel config requirements
```

### CODE REVIEW

File: Documentation/operations/system_requirements.rst [hunk 1]
```diff
@@ -82,7 +82,7 @@ RancherOS_                 >= 1.5.5
 .. _RancherOS: https://rancher.com/rancher-os/
 
 .. [#centos_foot] CentOS 7 requires a third-party kernel provided by `ElRepo <http://elrepo.org/tiki/tiki-index.php>`_
-    whereas CentOS 8 ships with a supported kernel.
+    whereas CentOS 8 ships with a supported kernel. Cilium features that require large BPF instruction set may not work with CentOS 7 because bpftool shipped with Cilium is not compatible with glibc version on CentOS 7.
```
Comment (julianwiedmann) [inline Documentation/operations/system_requirements.rst]:
```suggestion
    whereas CentOS 8 ships with a supported kernel. Cilium features that require large BPF programs (eg. VTEP integration) may not work on CentOS 7, as the ``bpftool` version shipped with Cilium can't probe for ``HAVE_LARGE_INSN_LIMIT`` support due to an incompatibility with the ``glibc`` version on CentOS 7.
```

Does that sound correct @vincentmli ? I'd rather not say "large BPF **instruction set**" when we actually mean the program size.

**Review by julianwiedmann — CHANGES_REQUESTED** [review]

File: Documentation/operations/system_requirements.rst (hunk 1 above)
Comment (vincentmli) [reply Documentation/operations/system_requirements.rst]: got it

**Comment by joestringer** [comment]
Cilium is distributed as a container with the glibc dependency, how does this issue arise?

**Comment by vincentmli** [comment]
> Cilium is distributed as a container with the glibc dependency, how does this issue arise?

ah, that is good point, an Engineer from https://github.com/F5Networks/k8s-bigip-ctlr is trying to validate Cilium VTEP integration with BIG-IP and trying to setup Cilium on CentOS 7, not sure why he tries CentOS 7 in the first place, it is not required by us, I guess this Cilium system requirements mentions CentOS 7 support. the engineer upgrade CentOS 7 kernel to 5.4 but still got this cilium agent log

```
level=info msg="clang (10.0.0) and kernel (5.4.206) versions: OK!" subsys=linux-datapath
level=info msg="linking environment: OK!" subsys=linux-datapath
level=fatal msg="VXLAN Tunnel Endpoint (VTEP) Integration: requires support for large programs (Linux 5.2.0 or newer)" subsys=linux-datapath
```
so my assumption 

1) CentOS 7 with 5.4 kernel does not have `HAVE_LARGE_INSN_LIMIT` enabled
or 
2) bpftool itself failed to probe CentOS 7 5.4 kernel for some reason

I thought is might be 2) because the engineer installed bpftool  package bpftool-3.10.0-1160.71.1.el7.x86_64 from  CentOS 7 repo and the `bpftool feature probe` get `null` result, then I copied bpftool shipped with Cilium and ran it on CentOS 7 host, I got glibc version not supported, this may leads me to wrong conclusion after thinking about what you said about bpftool in cilium agent container has glibc dependency. it might be the kernel not configured properly to support`HAVE_LARGE_INSN_LIMIT`? anyway, the engineer is not required to validate CentOS 7, and I don't want to spend too much time on digging CentOS 7, I had actually tried CentOS 7 while developing the VTEP feature, all sort of problem I don't remember clearly now :-). either we can drop this PR, or we could amend the footnote with some generic message mentioning CentOS 7 with upgraded kernel might still not operate well when cilium feature requires `HAVE_LARGE_INSN_LIMIT`, what do you think?

**Comment by joestringer** [comment]
Overall I think that anyone running on old CentOS is going to have to do a bit of extra legwork to ensure that they can get all the features they want, since CentOS already doesn't ship a new enough kernel by default and then the third party kernel could be configured in ways that don't support all of the features. That said, calling it out here for specific features like VTEP that are currently beta doesn't make much sense to me. I think we should simply define which kernel features are required for which advanced features (like we do already for some features). Then it's up to the distributions to come up with sane defaults, and the users of those distros to provide feedback to their distributor about which kernel features they expect to need.

As for copying the bpftool binary out from the Cilium container to run it, I think it's going too far to try to support such use cases from the binaries that are compiled into Cilium by default. AFAIK there's nothing special about the bpftool that Cilium ships in its container, so users on this platform should be able to just install any upstream bptool locally to use it.

**Comment by vincentmli** [comment]
I agree, I suspect not only VTEP, other advance features may not work on CentOS 7 by simply upgrading CentOS 7 kernel, I list VTEP just as example in the footnote, when I read the CentOS 7 footnote, it give me impression that if CentOS 7 user simply upgrade kernel, they are good to go, that is the impression I also get from the engineer who tried CentOS 7, the engineer actually also tried same kernel version on CentOS 8 as CentOS 7, CentOS 8 also works fine, it gives user impression it is not kernel version issue problem for users that does not know much about the relation between kernel and Cilium. so my idea is to not let the footnote mis-lead user to believe if they simply upgrade CentOS 7 kernel, everything will be supported, not sure how to phrase the sentence. with that said, I am also fine to drop the PR, no strong opinion, people can interpret phrase differently :-)

File: Documentation/operations/system_requirements.rst [hunk 2]
```diff
@@ -82,7 +82,7 @@ RancherOS_                 >= 1.5.5
 .. _RancherOS: https://rancher.com/rancher-os/
 
 .. [#centos_foot] CentOS 7 requires a third-party kernel provided by `ElRepo <http://elrepo.org/tiki/tiki-index.php>`_
-    whereas CentOS 8 ships with a supported kernel.
+    whereas CentOS 8 ships with a supported kernel. Cilium features that require large BPF programs (eg. VTEP integration) may not work on CentOS 7, as the ``bpftool`` version shipped with Cilium can't probe for ``HAVE_LARGE_INSN_LIMIT`` support due to an incompatibility with the ``glibc`` version on CentOS 7.
```
Comment (joestringer) [inline Documentation/operations/system_requirements.rst]: Maybe we can just make a more generic message and improve the base kernel requirements statements below.

```suggestion
    whereas CentOS 8 ships with a supported kernel. Note that some more advanced features may not be available on Centos 7 even with a third-party kernel. For full details on which kernel config options must be enabled in order to use various features, see the section on :ref:`admin_kernel_version` requirements below.
```

**Comment by joestringer** [comment]
If we update this PR something like the above, then please also remove/update the release note in the description above. At this point we appear to be trying to better communicate kernel support in CentOS kernels and not the glibc thing with bpftool.

**Review by joestringer — APPROVED** [review]
LGTM. If there's any other additional changes to suggest to the kernel requirements section, feel free to follow up in a subsequent PR.

**Review by julianwiedmann — APPROVED** [review]

**Pull request merged**